### PR TITLE
feat: add  environment flow sync process

### PR DESCRIPTION
**Pull request by ytvnr**
## Issue

https://gravitee.atlassian.net/browse/APIM-5624

## Description

Create sync process for Environment Flows
This PR contains:
- a first version of the environment flow model
- repository sync
- distributed sync

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-sncttqqmii.chromatic.com)

### CODE REVIEW

File: gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/environmentflow/EnvironmentFlow.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    private List<Step> policies;
+
+    @JsonIgnore
+    public List<Plugin> getPlugins() {
+        return this.policies == null
+            ? List.of()
+            : this.policies.stream().filter(Step::isEnabled).map(Step::getPlugins).flatMap(List::stream).toList();
+    }
+
+    public enum Phase {
```
Comment (ThibaudAV) [inline gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/environmentflow/EnvironmentFlow.java:74]: don't we already have this enum in the definition?

File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/EnvironmentFlow.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+     */
+    private Date createdAt;
+    /**
+     * The Api last updated date
+     */
+    private Date updatedAt;
+    /**
+     * The current runtime life cycle state.
+     */
+    private LifecycleState lifecycleState;
```
Comment (ThibaudAV) [inline gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/EnvironmentFlow.java]: Just a preference: 
I wouldn't keep a dependency here. I find that the state of an env-flow is its own. It shouldn't depend on anything else and should be free to evolve without dependency.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml [hunk 3]
```diff
@@ -5985,6 +5985,8 @@ components:
                 - UNPUBLISH_API
                 - UNPUBLISH_API_RESULT
                 - UNPUBLISH_DICTIONARY
+                - START_ENVIRONMENT_FLOW
+                - STOP_ENVIRONMENT_FLOW
```
Comment (ThibaudAV) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml]: same just a preference: 😁 
I'd use the same "jargon" as the product. if it's called a deploy and undeploy, I'd add it here. 
start stop makes it look like there's something spinning in the background 🤷‍♂️

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/repository/synchronizer/environmentflow/EnvironmentFlowSynchronizer.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            // fetch per page
+            .flatMap(events ->
+                Flowable
+                    .just(events)
+                    .doOnNext(e -> log.debug("New environment flow events fetch"))
+                    .flatMapIterable(e -> e)
+                    .groupBy(Event::getType)
+                    .flatMap(eventsByType -> {
+                        if (eventsByType.getKey() == EventType.START_ENVIRONMENT_FLOW) {
+                            return prepareForDeployment(eventsByType);
```
Comment (ThibaudAV) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/repository/synchronizer/environmentflow/EnvironmentFlowSynchronizer.java:113]: ❤️ Deploy 😁

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/common/deployer/EnvironmentFlowDeployer.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    private final EnvironmentFlowManager environmentFlowManager;
+    private final DistributedSyncService distributedSyncService;
+
+    @Override
+    public Completable deploy(final EnvironmentFlowReactorDeployable deployable) {
+        return Completable.fromRunnable(() -> {
+            ReactableEnvironmentFlow reactableEnvironmentFlow = deployable.reactableEnvironmentFlow();
+            try {
+                environmentFlowManager.register(reactableEnvironmentFlow);
+                log.debug("Environment Flow [{}] deployed ", deployable.environmentFlowId());
```
Comment (ThibaudAV) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/common/deployer/EnvironmentFlowDeployer.java:44]: maybe add id + version ? to identify it properly 🤔

**Review by ThibaudAV — APPROVED** [review]
Too good . can't wait to play with it

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/common/deployer/EnvironmentFlowDeployer.java (hunk 5 above)
Comment (ThibaudAV) [reply gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/common/deployer/EnvironmentFlowDeployer.java:44]: same question everywhere to idetify it easily. but maybe useless

File: gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/environmentflow/EnvironmentFlow.java (hunk 1 above)
Comment (ytvnr) [reply gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/environmentflow/EnvironmentFlow.java:74]: I did not find anything that fit perfectly 😞

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/common/deployer/EnvironmentFlowDeployer.java (hunk 5 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/common/deployer/EnvironmentFlowDeployer.java:44]: I did it in this class only, I think it is enough

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml (hunk 3 above)
Comment (ytvnr) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml]: renamed to DEPLOY_ENVIRONMENT_FLOW

File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/EnvironmentFlow.java (hunk 2 above)
Comment (ytvnr) [reply gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/EnvironmentFlow.java]: created a dedicated object :)

**Pull request merged**